### PR TITLE
fixes #3377 - add altlinux support

### PR DESCRIPTION
**Pull request by sergelogvinov**
Ticket http://projects.theforeman.org/issues/3377

Add altlinux os support.

### CODE REVIEW

**Comment by ohadlevy** [comment]
[test]

**Comment by ohadlevy** [comment]
@GregSutcliffe can you review please?

File: app/models/operatingsystems/altlinux.rb [hunk 1]
```diff
@@ -0,0 +1,39 @@
+class Altlinux < Operatingsystem
+
+  #PXEFILES = {:kernel => "vmlinuz", :initrd => "full.cz", :stagename => "altinst"}
```
Comment (GregSutcliffe) [inline app/models/operatingsystems/altlinux.rb]: Nitpick: leftover comment?

File: app/models/operatingsystems/altlinux.rb [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+  def class
+    Operatingsystem
+  end
+
+  def boot_files_uri(medium, architecture)
+    raise ::Foreman::Exception.new(N_("invalid medium for %s"), to_s) unless media.include?(medium)
+    raise ::Foreman::Exception.new(N_("invalid architecture for %s"), to_s) unless architectures.include?(architecture)
+
+    PXEFILES.values.collect do |img|
+      if img == 'altinst'
```
Comment (GregSutcliffe) [inline app/models/operatingsystems/altlinux.rb]: Does this if get used at all? It seems the "altinst" parameter has been commented out...

File: config/routes.rb [hunk 3]
```diff
@@ -292,6 +292,8 @@
   match 'status', :to => 'home#status', :as => "status"
 
   # match for all unattended scripts
+  match 'unattended/(:action/(:id(.scm)))',    :controller => 'unattended', :format => false
+  match 'unattended/(:action/(:id(.tar)))',    :controller => 'unattended', :format => false
```
Comment (GregSutcliffe) [inline config/routes.rb]: Why would the install be trying to get .tar or .scm files from Foreman?

**Comment by GregSutcliffe** [comment]
@sergelogvinov a few small questions, but looks ok overall. I feel it might need some tests though - @domcleal thoughts?

File: app/views/unattended/alterator.erb [hunk 4]
```diff
@@ -0,0 +1,23 @@
+<% if @metadata.to_s == 'vm-profile' -%>
```
Comment (domcleal) [inline app/views/unattended/alterator.erb]: I don't recognise `@metadata`, is that one of ours?

Comment (domcleal) [inline app/views/unattended/alterator.erb]: Ignore me, I see you added it :)

File: config/routes.rb (hunk 3 above)
Comment (sergelogvinov) [inline config/routes.rb]: Alterator need pkg-groups.tar file. Foreman use as proxy server and get this file from repository #{@mediapath}/Metadata/pkg-groups.tar

scm its just text file.

File: config/routes.rb [hunk 5]
```diff
@@ -291,6 +291,9 @@
   match 'statistics', :to => 'statistics#index', :as => "statistics"
   match 'status', :to => 'home#status', :as => "status"
 
+  # match only for alterator unattended scripts
+  match 'unattended/provision/:metadata',  :controller => 'unattended', :action => 'provision', :format => false,
```
Comment (ohadlevy) [inline config/routes.rb]: why format: false? isnt it html/xml/json etc?

File: config/routes.rb [hunk 6]
```diff
@@ -291,6 +291,9 @@
   match 'statistics', :to => 'statistics#index', :as => "statistics"
   match 'status', :to => 'home#status', :as => "status"
 
+  # match only for alterator unattended scripts
+  match 'unattended/provision/:metadata',  :controller => 'unattended', :action => 'provision', :format => false, 
+    :constraints => { :metadata => /(autoinstall\.scm|vm-profile\.scm|pkg-groups\.tar)/ }
```
Comment (ohadlevy) [inline config/routes.rb:337]: I am not familiar with altlinux, but are those mandatory or can they be provided via a snippet etc or something similar?
alternatively,  does it make more sense to define a bare minimal settings (such as limited groups) and then add more when needed in a template?

File: app/views/unattended/alterator.erb [hunk 7]
```diff
@@ -0,0 +1,23 @@
+<% if @metadata.to_s == 'vm-profile.scm' -%>
+<%= @host.diskLayout %>
+<% elsif @metadata.to_s == 'pkg-groups.tar' -%>
+<%= media_get_content("#{@mediapath}/Metadata/pkg-groups.tar") %>
```
Comment (ohadlevy) [inline app/views/unattended/alterator.erb]: this does not make any sense to me, why are you asking foreman to make another web request to itself?

you can simply do

```
<%= snippet('pkg-groups') %>
```

and avoid the whole roundtrip

File: config/routes.rb (hunk 6 above)
Comment (sergelogvinov) [inline config/routes.rb:337]: $ tar -tvf pkg-groups.tar  | wc -l
92

pkg-groups specific on version distributive and not needed customization. However, you can set specific pkg list/group to install on autoinstall.scm (section /pkg-install) Its like software patterns on Yast or packages on Kickstart. 
As solution i use media_get_content() for delivery this file.

use snippet how, what for?

Comment (ohadlevy) [inline config/routes.rb:337]: as i understand it, what you are doing is:
1. request unattended tempate.
2. as foreman generate the request, it parse the template, and see there is media_get_content method which make an HTTP call to foreman again, in order to include that content.

I would assume that 2 can be replaced with content which is stored as a snippet, and we don't need all of the extra roundtrips?

Comment (sergelogvinov) [inline config/routes.rb:337]: Ok. I do not know how stored in a snippet this structure:

$tar -tvf pkg-groups.tar
drwxr-xr-x root/root         0 2013-09-06 10:45 groups/
-rw-r--r-- 598/599         245 2010-08-23 13:07 groups/dhcp-server-a.directory
-rw-r--r-- 598/599         207 2010-08-23 13:07 groups/centaurus-gnome.directory
-rw-r--r-- 598/599         196 2010-08-23 13:07 groups/buildsystem.directory
...
drwxr-xr-x root/root         0 2013-09-06 21:29 lists/
-rw-r--r-- 598/599           6 2011-04-12 11:29 lists/gnome-phone
-rw-r--r-- 598/599          10 2011-04-12 11:29 lists/eclipse
-rw-r--r-- 598/599         123 2011-04-12 11:29 lists/gnome-office
...

any ideas?

**Comment by ohadlevy** [comment]
@sergelogvinov ping ?

**Comment by sergelogvinov** [comment]
https://github.com/theforeman/foreman/pull/966#discussion_r7191742

**Comment by ohadlevy** [comment]
@sergelogvinov ah, somehow github hides that thread...

so it actually expects a tar with a directory structure?!! any alternatives?

**Comment by sergelogvinov** [comment]
group in pkg-groups.tar use in autoinstall.scm (section /pkg-install).

alternatives..
1. set proxy on apache/nginx/lighttpd before foreman (not universal method, pkg-groups.tar always one version)
2. refactoring alterator-pxe installer - i am not maintainer altlinux
3. use proxy on ruby, media_get_content()

I use media_get_content(), because this very simple method. 

what do you think? ideas?

**Comment by sergelogvinov** [comment]
Maybe create tar file from memory use code like https://gist.github.com/sinisterchipmunk/1335041 (see tar function).
Body get from snippet, as result we have tar file with one file inside (section /pkg-install in autoinstall.scm anymore not use). 

But how to implement it in Foreman style.

**Comment by ohadlevy** [comment]
@sergelogvinov sorry for the long pause in review, hopefully we can get it... questions:
1. I believe i still didnt get the whole picture, what is the relationship to a proxy?
2. if we must provide it as a tar, i assume the tar is static, and as such, we could put it under public, then you could simply File.read(Rails.root + /public...) instead of a second http call.

TBH I don't understand whats the idea behind the tar, can you refer me to any docs?

**Comment by dLobatog** [comment]
@sergelogvinov  ping, can you rebase?

**Comment by sergelogvinov** [comment]
Sorry for the delay.
Give me two days please.

File: app/models/operatingsystem.rb [hunk 8]
```diff
@@ -52,6 +52,7 @@ class Operatingsystem < ActiveRecord::Base
                'Redhat'    => %r{RedHat|Centos|Fedora|Scientific|SLC|OracleLinux}i,
                'Suse'      => %r{OpenSuSE|SLES|SLED}i,
                'Windows'   => %r{Windows}i,
+               'Altlinux' => %r{Altlinux}i,
```
Comment (dLobatog) [inline app/models/operatingsystem.rb]: Please align the `=>` sign

File: app/models/operatingsystems/altlinux.rb [hunk 9]
```diff
@@ -0,0 +1,34 @@
+class Altlinux < Operatingsystem
+
+  PXEFILES = {:kernel => "vmlinuz", :initrd => "full.cz" }
+
+  def mediumpath host
+    medium_uri(host).to_s
+  end
+
+  def class
+    Operatingsystem
```
Comment (dLobatog) [inline app/models/operatingsystems/altlinux.rb:6]: Why do you need to redefine this?

File: lib/foreman/renderer.rb [hunk 10]
```diff
@@ -24,6 +24,10 @@ def render_safe template, allowed_methods = [], allowed_vars = {}
       end
     end
 
+    def media_get_content(url)
```
Comment (dLobatog) [inline lib/foreman/renderer.rb]: As @ohadlevy  mentions above, this should be unnecessary.

**Comment by dLobatog** [comment]
:+1: to most of it, please use `snippet` instead of media_get_content, reply to the comments, and most likely this is in. 

Thank you @sergelogvinov  :smile:

File: app/models/operatingsystems/altlinux.rb (hunk 9 above)
Comment (domcleal) [inline app/models/operatingsystems/altlinux.rb:6]: Change in style since PR was written, no longer required.

**Comment by domcleal** [comment]
A lot's changed since this was opened, I hope you're serious about including it if asking for updates/rebases.  Please note that templates have to go via community-templates now, don't merge directly into the Foreman tree.

**Comment by sergelogvinov** [comment]
Pull request altlinux templates https://github.com/theforeman/community-templates/pull/86

**Comment by domcleal** [comment]
community-templates PR was merged, thanks [test].  back to @eLobato

**Comment by lzap** [comment]
In regard to the logo, I am unable to find license under the AltLinux logo author is providing it.

http://www.altlinux.org/%D0%9B%D0%BE%D0%B3%D0%BE%D1%82%D0%B8%D0%BF%D1%8B

To me it looks like we need to either contact them or find the license, or remove the logo and replace it with something from our dummy icons repo: 

https://github.com/lzap/foreman-dummy-icons/blob/master/21x21-goldenrod/a.png

**Comment by dLobatog** [comment]
@sergelogvinov thank you! It looks nearly ready to merge now, there is one test that is broken after the latest update

https://github.com/theforeman/foreman/blob/develop/test/unit/operatingsystem_test.rb#L178 , those arrays need to include 'Altlinux' too.

Also let's remove the logo for the moment as it's unclear whether or not it can be used here, when we figure out if it can be used we'll put it and we can use one of @lzap 's logos from this project for the moment, https://github.com/lzap/foreman-dummy-icons , black is my preference.

**Comment by lzap** [comment]
I've already added the icon in this PR:

https://github.com/theforeman/foreman/pull/1558/files

Please remove it from here. Discuss there.

LZ

## 

Later,
 Lukas "lzap" Zapletal

**Comment by dLobatog** [comment]
[test] , thanks for pushing new changes

File: test/unit/operatingsystem_test.rb [hunk 11]
```diff
@@ -176,8 +176,8 @@ class OperatingsystemTest < ActiveSupport::TestCase
 
     test "families_as_collection contains correct names and values" do
       families = Operatingsystem.families_as_collection
-      assert_equal ["AIX", "Arch Linux", "Debian", "FreeBSD", "Gentoo", "Junos", "Red Hat", "SUSE", "Solaris", "Windows"], families.map(&:name).sort
-      assert_equal ["AIX", "Archlinux", "Debian", "Freebsd", "Gentoo", "Junos", "Redhat", "Solaris", "Suse", "Windows"], families.map(&:value).sort
+      assert_equal ["AIX", "ALTLinux", "Arch Linux", "Debian", "FreeBSD", "Gentoo", "Junos", "Red Hat", "SUSE", "Solaris", "Windows"], families.map(&:name).sort
```
Comment (dLobatog) [inline test/unit/operatingsystem_test.rb]: As you put 'Altlinux' [there](https://github.com/theforeman/foreman/pull/966/files#diff-d4ad9c79d567051cceead5ad8129c048R22) , it should be the same here. Note only the first letter is upper case. 

It's the only test failing, sorry for going back and forth on this. Mention me in a comment so I notice via email that you updated this so I can take a look at it quickly.
http://ci.theforeman.org/job/test_develop_pr_core/1599/database=mysql,puppet=3.0,ruby=1.9.3/testReport/junit/(root)/OperatingsystemTest__families/test_0008_families_as_collection_contains_correct_names_and_values/

**Comment by dLobatog** [comment]
@sergelogvinov the test is still failing as explained above, it's a really minor thing but breaks the suite, let me know when it's fixed. Thanks!

**Comment by dLobatog** [comment]
[test]

File: lib/foreman/renderer.rb [hunk 12]
```diff
@@ -8,7 +8,7 @@ module Renderer
 			:multiboot, :jumpstart_path, :install_path, :miniroot,
 			:media_path, :param_true?, :param_false? ]
 
-    ALLOWED_VARIABLES = [ :arch, :host, :osver, :mediapath, :static,
+    ALLOWED_VARIABLES = [ :arch, :host, :osver, :mediapath, :madiaserver, :static,
```
Comment (dLobatog) [inline lib/foreman/renderer.rb:11]: madiaserver -> mediaserver

**Comment by dLobatog** [comment]
@sergelogvinov I'm not sure what you did on the latest commit aside from fixing the OS test, it seems to have gone back in time, and it unsquashed all the commits. 

Remember templates are already merged in community-templates here: https://github.com/theforeman/community-templates/pull/86/

The Altlinux logo (app/assets/images/Altlinux.png) cannot be included either.

You can see the failed tests here, http://ci.theforeman.org/job/test_develop_pr_core/1603/#showFailuresLink , I think it lacks some permission for the route 'unattended/provision' and removing the templates.

I'm sure you can get back to the point where it was about to get merged by playing around with git a little :smile:

**Comment by dLobatog** [comment]
[test]

**Comment by dLobatog** [comment]
@sergelogvinov I was able to recover your branch properly, since it seemed like you did a merge commit, and squashed it. 

Merged as https://github.com/theforeman/foreman/commit/22671c505ea10d7896b8590c4a88ff82c2476ba3 , thanks @sergelogvinov and Kot!

**Pull request closed without merging**